profiling/server.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package profiling
+
+import (
+	"errors"
+	"net/http"
+	"net/http/pprof"
+	"strconv"
+	"sync"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// profilingPort is the port where we expose profiling information if profiling is enabled
+	profilingPort = ":8008"
+
+	// profilingKey is the name of the key in config-observability config map that indicates whether profiling
+	// is enabled of disabled
+	profilingKey = "profiling.enable"
+)
+
+// Handler holds the main HTTP handler and a flag indicating
+// whether the handler is active
+type Handler struct {
+	enabled    bool
+	enabledMux sync.Mutex
+	handler    http.Handler
+	log        *zap.SugaredLogger
+}
+
+// NewHandler create a new ProfilingHandler which serves runtime profiling data
+// according to the given context path
+func NewHandler(logger *zap.SugaredLogger, configMap *corev1.ConfigMap) *Handler {
+	const pprofPrefix = "/debug/pprof/"
+
+	mux := http.NewServeMux()
+	mux.HandleFunc(pprofPrefix, pprof.Index)
+	mux.HandleFunc(pprofPrefix+"cmdline", pprof.Cmdline)
+	mux.HandleFunc(pprofPrefix+"profile", pprof.Profile)
+	mux.HandleFunc(pprofPrefix+"symbol", pprof.Symbol)
+	mux.HandleFunc(pprofPrefix+"trace", pprof.Trace)
+
+	profilingEnabled, _ := readProfilingFlag(configMap)
+
+	logger.Infof("Profiling enabled: %t", profilingEnabled)
+
+	return &Handler{
+		enabled: profilingEnabled,
+		handler: mux,
+		log:     logger,
+	}
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.enabledMux.Lock()
+	defer h.enabledMux.Unlock()
+	if h.enabled {
+		h.handler.ServeHTTP(w, r)
+	} else {
+		http.NotFoundHandler().ServeHTTP(w, r)
+	}
+}
+
+func readProfilingFlag(configMap *corev1.ConfigMap) (bool, error) {
+	profiling, ok := configMap.Data[profilingKey]
+	if !ok {
+		return false, nil
+	}
+	enabled, err := strconv.ParseBool(profiling)
+	if err != nil {
+		return false, errors.New("Failed to parse the profiling flag")
```

Comment:
This should wrap the original error instead of generating a new one. Import `github.com/pkg/errors` and use `errors.Wrapf(err, "failed to parse the profiling flag")`. Always start error messages lower-cased as per the Golang code guide.